sklego/meta/grouped_transformer.py

```diff
@@ -0,0 +1,159 @@
+        self.transformer = transformer
+        self.groups = groups
+        self.use_global_model = use_global_model
+
+    def __check_value_columns(self, X):
+        """Do basic checks on the value columns"""
+        if isinstance(X, pd.DataFrame):
+            X_value = check_array(X.drop(columns=self.groups))
+        else:
+            # If not a df, we impose standard checks also on grouping columns
+            X = check_array(X)
+            X_value = np.delete(X, as_list(self.groups), axis=1)
+
+        return X_value
+
+    def __check_grouping_columns(self, X):
+        """Do basic checks on grouping columns"""
+        missings = False
+
+        if isinstance(X, pd.DataFrame):
+            X_group = X.loc[:, as_list(self.groups)]
+            if X_group.isnull().sum().sum() > 0:
+                missings = True
+        else:
+            X_group = check_array(X)[:, as_list(self.groups)]
+            if np.isnan(X_group).sum(axis=None) > 0:
+                missings = True
+
+        if missings:
+            raise ValueError("X has NaN values")
+
+        # The grouping part we always want as a DataFrame with range index
+        return pd.DataFrame(X_group).reset_index(drop=True)
+
+    def __fit_single_group(self, group, X, y=None):
+        try:
+            return clone(self.transformer).fit(X, y)
+        except Exception as e:
+            raise type(e)(f"Exception for group {group}: {e}")
+
+    def __fit_grouped_transformer(
+        self, X_group: pd.DataFrame, X_value: np.array, y=None
+    ):
+        """Fit a transformer to each group"""
+        # Make the groups based on the groups dataframe, use the indices on the values array
+        group_indices = X_group.groupby(X_group.columns.tolist()).indices
+
+        if y:
+            grouped_transformers = {
+                # Fit a clone of the transformer to each group
+                group: self.__fit_single_group(group, X_value[indices, :], y[indices])
+                for group, indices in group_indices.items()
+            }
+        else:
+            grouped_transformers = {
+                group: self.__fit_single_group(group, X_value[indices, :])
+                for group, indices in group_indices.items()
+            }
+
+        return grouped_transformers
+
+    def __check_transformer(self):
+        if not hasattr(self.transformer, "transform"):
+            raise ValueError(
+                "The supplied transformer should have a 'transform' method"
+            )
+
+    def fit(self, X, y=None):
+        """
+        Fit the transformers to the groups in X
+
+        :param X: Array-like with at least two columns, of which at least one corresponds to groups defined in init,
+                  and the remaining columns represent the values to transform.
+        :param y: (Optional) target variable
+        """
+        self.__check_transformer()
+
+        X_value = self.__check_value_columns(X)
```

Comment:
Yep. Found a bug here.

```python
from sklego.datasets import load_penguins
from sklego.meta import GroupedEstimator
from sklearn.preprocessing import StandardScaler

X = load_penguins()['data']
meta = GroupedTransformer(StandardScaler(), groups=[0])
meta.fit_transform(X[:, :3])
```

If you run this you'll get an error. Here's the context, this is what `X[:, :3]` looks like;

```python
array([['Torgersen', 39.1, 18.7],
       ['Torgersen', 39.5, 17.4],
       ['Torgersen', 40.3, 18.0],
       ...,
       ['Dream', 49.6, 18.2],
       ['Dream', 50.8, 19.0],
       ['Dream', 50.2, 18.7]], dtype=object)
```

You'll notice that column 0 (the one I want to group over) is still being checked as if it isn't a variable that I am grouping over. In this case, grouping over a string should be fine but I'm getting a 

```
ValueError: could not convert string to float: 'Torgersen'
```

Reply:
This one was terrible. Didn't know about arrays with mixed types, but I'll steer clear of them as much as possible. 
I don't think my solution is very nice, so if you have ideas for improvements let me know!

Reply:
I think there's still a similar bug. 

```python
from sklego.datasets import load_penguins
from sklego.meta import GroupedTransformer, EstimatorTransformer
from sklearn.dummy import DummyRegressor
from sklearn.preprocessing import StandardScaler

tfm = EstimatorTransformer(StandardScaler(), predict_func='predict')
meta = GroupedTransformer(StandardScaler(), groups=[0, -1])

df = load_penguins(as_frame=True).dropna()
X = df.drop(columns='species').values
```

This is what the dataset looks like; 

```python
array([['Torgersen', 39.1, 18.7, 181.0, 3750.0, 'male'],
       ['Torgersen', 39.5, 17.4, 186.0, 3800.0, 'female'],
       ['Torgersen', 40.3, 18.0, 195.0, 3250.0, 'female'],
       ...,
       ['Dream', 49.6, 18.2, 193.0, 3775.0, 'male'],
       ['Dream', 50.8, 19.0, 210.0, 4100.0, 'male'],
       ['Dream', 50.2, 18.7, 198.0, 3775.0, 'female']], dtype=object)
```

Alas, we still get an error. 

```python
meta.fit_transform(X)
# ValueError: could not convert string to float: 'male'
```

Might be good to add this as a unit-test. It seems to work fine if `groups=[0, 5]` but it fails when `groups=[0, -1]`.

Reply:
I've found another interesting edge case. 

```
from sklego.datasets import load_penguins
from sklego.meta import GroupedTransformer, EstimatorTransformer
from sklearn.dummy import DummyRegressor
from sklearn.preprocessing import StandardScaler

tfm = EstimatorTransformer(StandardScaler(), predict_func='predict')
meta = GroupedTransformer(StandardScaler(), groups=[0, 5])

df = load_penguins(as_frame=True).dropna()
X = df.drop(columns='species').values
meta.fit_transform(X[:, 3])
```

It gives a reshape error instead of a "the dataframe you pass in has three columns, we can't select column 5".

Reply:
The first one seems to have been resolved by the last commit(`Grouping on string columns in numpy arrays`, because I don't have this error. I will implement a test though for this specific case.

Reply:
The other edge case is a typo (I think?), you have `X[:, 3]`, so selecting only the third column. If I do `X[:, :3]` it does raise an IndexError